project-minimization-plugin/src/main/kotlin/org/plan/research/minimization/plugin/services/BenchmarkService.kt

```diff
@@ -0,0 +1,148 @@
+
+import kotlin.coroutines.resume
+import kotlin.io.path.Path
+import kotlin.io.path.readText
+import kotlinx.coroutines.*
+
+@Service(Service.Level.PROJECT)
+class BenchmarkService(private val rootProject: Project, private val cs: CoroutineScope) {
+    fun benchmark() = cs.launch {
+        val config = readConfig().getOrNull()
+        config ?: run {
+            return@launch
+        }
+        val filteredProjects = config
+            .projects
+            .filter { it.isSuitableForGradleBenchmarking(allowAndroid = false) }  // FIXME
+        withBackgroundProgress(rootProject, "Running Minimization Benchmark") {
+            reportSequentialProgress(filteredProjects.size) { reporter ->
+                filteredProjects.forEach { project ->
+                    reporter.itemStep("Minimizing ${project.name}") {
+                        catch({
+                            project.process()
+                        },
+                        ) {
+                            }
+                    }
+                }
+            }
+        }
+    }
+
+    private suspend fun closeProject(project: Project) = withContext(Dispatchers.EDT) {
+        ProjectManager.getInstance().closeAndDispose(project)
+    }
+
+    private suspend fun readConfig(): Option<BenchmarkConfig> = option {
+        val projectRoot = rootProject.guessProjectDir() ?: raise(None)
+        val configurationFile = projectRoot.findFile("config.yaml") ?: raise(None)
+        val content = withContext(Dispatchers.IO) { configurationFile.readText() }
+        Yaml.default.decodeFromString(BenchmarkConfig.serializer(), content)
+    }
+
+    private fun getBenchmarkProjectRoot(project: BenchmarkProject): Option<Path> = option {
+        val root = rootProject.guessProjectDir()?.toNioPath() ?: raise(None)
+        root.resolve(Path(project.path))
+    }
+
+    private suspend fun openBenchmarkProject(project: BenchmarkProject): Option<Project> = option {
+        val root = getBenchmarkProjectRoot(project).bind()
+        try {
+            val project = service<ProjectOpeningService>().openProject(root)
+            ensureNotNull(project)
+        } catch (e: Throwable) {
+            e.printStackTrace()
+            raise(None)
+        }
+    }
+
+    private suspend fun BenchmarkProject.process(): Project? {
+        // Either<MinimizationError, Project>?
+        val gradleBuildTask = loadReproduceScript(this@process) ?: "build"
+        val openedProject = openBenchmarkProject(this@process).getOrNull() ?: return null
+
+        if (openedProject.isDisposed) {
+            throw IllegalStateException("Project is already disposed")
+        }
+
+        try {
+            setMinimizationSettings(openedProject, this.settingsFile)
+            val minimizationService = openedProject.service<MinimizationService>()
+
+            val result = suspendCancellableCoroutine { cont ->
+                minimizationService.minimizeProject { context -> cont.resume(context) }
```

Comment:
I see now that there are no problems with using the private method `minimizeProjectImpl`. I had some concerns about background progress, but we already have some here. So what you can do:
1. make `minimizeProjectImpl` public
2. rename `minimizeProjectImpl` to `minimizeProjectAsync`
3. get rid of suspendCancellableCoroutine